### PR TITLE
Search promotions module

### PR DESCRIPTION
**Pull request by kaedroho**
Moves editors picks out of ``wagtailsearch`` into a new module called ``contrib.wagtailsearchpicks``.

This makes "Promoted search results" optional and also resolves the two-way dependency between ``wagtailsearch`` and ``wagtailcore``

TODO:
 - [x] Move ``EditorsPick`` model
 - [x] Add ``{% get_search_picks %}`` template tag
 - ~~Move locale files~~ (best to do this just before/after merge)
 - [x] Documentation

### CODE REVIEW

**Comment by davecranwell** [comment]
Is this an opportunity to finally rid the code of references to  "Editor's picks"? The term has already been agreed as confusing and only exists in a few places where it was harder to remove. It's ostensibly known as "Promoted search results".

If a short name is desirable for some situations "Promoted search" is only 2 chars longer than "Editors picks"  and would at least unify the names somewhat.

**Comment by kaedroho** [comment]
Ready for review

**Comment by davecranwell** [comment]
Sorry to return to this, but are we really going for `wagtailsearchpicks` over `wagtailpromotedsearch` or `wagtailsearchpromo` or similar? Continuing this mixture of terminology "picks" vs "promoted" seems undesirable if we've finally got a chance to move to something uniform.

**Comment by davecranwell** [comment]
Agreed to use "wagtailsearchpromotions".

**Comment by kaedroho** [comment]
- Renamed ``SearchPick`` model to ``SearchPromotion``
- Renamed ``wagtailsearchpicks`` app to ``wagtailsearchpromotions``

File: docs/reference/contrib/searchpromotions.rst [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+The ``searchpromotions`` module is not enabled by default. To install it, add ``wagtail.contrib.wagtailsearchpromotions`` to ``INSTALLED_APPS`` in your project's Django settings file.
+
+
+.. code-block:: python
+
+    INSTALLED_APPS = [
+        ...
+
+        'wagtail.contrib.wagtailsearchsearchpromotions',
```
Comment (gasman) [inline docs/reference/contrib/searchpromotions.rst]: typo here - duplication of 'search'

Comment (kaedroho) [inline docs/reference/contrib/searchpromotions.rst]: Oops...

File: docs/topics/search/searching.rst [hunk 2]
```diff
@@ -100,28 +100,12 @@ And here's a template to go with it:
...
-
-
-Each editors pick contains the following fields:
-
-  ``page``
-    The page object associated with the pick. Use ``{% pageurl editors_pick.page %}`` to generate a URL or provide other properties of the page object.
-
-  ``description``
-    The description entered when choosing the pick, perhaps explaining why the page is relevant to the search terms.
+This functionality is provided by the :mod:`~wagtail.contrib.wagtailsearchpicks` contrib module.
```
Comment (gasman) [inline docs/topics/search/searching.rst]: Should be `wagtail.contrib.wagtailsearchpromotions`. (Would be good to have a link to the relevant docs here - or will that happen automatically once the name is fixed?)

Comment (kaedroho) [inline docs/topics/search/searching.rst]: yeah, ``:mod:wagtail.contrib.wagtailsearchpromotions`` links to ``..module:: wagtail.contrib.wagtailsearchpromotions``

**Comment by gasman** [comment]
Looks good!

* I see there are several places that still refer to "Editor's Picks" / "search picks", some more visible/important to fix than others - for example, the success message on add/edit, and the related_name on the SearchPromotion model.
* The migrations worked well when I tried them, and successfully preserved existing search pick data - but I see the logic in there is a bit fiddly. It would be good to have some comments in there to explain what's going on, to make me more confident that it's going to work in all situations (for example, running the `wagtailsearch.0003_remove_editors_pick` migration before the new module has been added to INSTALLED_APPS).
* Please could you write an 'upgrade considerations' item for the release notes, explaining what existing site owners need to do?

**Comment by kaedroho** [comment]
> I see there are several places that still refer to "Editor's Picks" / "search picks", some more visible/important to fix than others - for example, the success message on add/edit, and the related_name on the SearchPromotion model.

I left the UI as it is as it's wrong on master anyway (was planning on doing this in a seprate PR).

Since we're no longer recommending users access promoted results using the ``Query`` object, I didn't see much benefit to changing the related name (and most implementations using promoted results will be using this related name).

**Comment by kaedroho** [comment]
Done!

**Comment by kaedroho** [comment]
Rebased

**Comment by SalahAdDin** [comment]
:+1:  Man, don't forget change de name for wagtailssearchpromotions to searchpromotions.

**Comment by kaedroho** [comment]
Removed the upgrade consideration commit (so I don't have to keep rebasing this branch). You can fing the commit for it here: https://github.com/kaedroho/wagtail/commit/1f4a8ec79c3d10834959e918ff3404e0146da452

**Comment by kaedroho** [comment]
Happy for me to merge this? I got a couple of changes to make that depend on this

**Comment by gasman** [comment]
Merged.

> Removed the upgrade consideration commit (so I don't have to keep rebasing this branch). You can find the commit for it here: kaedroho@1f4a8ec

For future reference, please don't do this. Fixing merge conflicts that arise while a PR is awaiting review is my responsibility as the reviewer, and it's easier to do that when it's part of the PR than when it's an orphaned commit on someone else's fork... :-)

**Pull request merged**